### ISSUE TITLE
Object Storage: Download object and write to file

### ISSUE THREAD

**Issue by lsarni** [ContainX/openstack4j#918]
I'm using this to download an object and write it to a file:
client.objectStorage().objects().download(publicBucket, externalFileName).writeToFile(new File(localFile));

I was expecting it to fail with a StorageException if the object didn't exist but this is not the case, it creates a file of 70 bytes.
Is this the expected outcome? I was expecting the http response to be not found and that to be in the storage exception.

**Comment by auhlig** [comment]
Mh. At first glance yes. Could you check the response from Swift, @lsarni ?

**Comment by lsarni** [comment]
Using [this](http://developer.openstack.org/api-ref/object-storage/?expanded=create-or-update-object-metadata-detail,get-object-content-and-metadata-detail#get-object-content-and-metadata) I get:

```
HTTP/1.1 404 Not Found
Content-Length: 70
Content-Type: text/html; charset=UTF-8
X-Trans-Id: tx0b057418fd4d4de2aac6c-0058777297
Date: Thu, 12 Jan 2017 12:12:07 GMT

<html><h1>Not Found</h1><p>The resource could not be found.</p></html>
```

I also tried it on a console app, outside my main project, and it still downloads a file.

**Comment by auhlig** [comment]
Looking at the content lenght, I guess the response is written to the file?

**Comment by lsarni** [comment]
Yes, I was trying to download an image but if I download a txt the content is:

<html><h1>Not Found</h1><p>The resource could not be found.</p></html>

**Comment by auhlig** [comment]
Mh. It should return null, as the download() method leads to get(), looking at https://github.com/ContainX/openstack4j/blob/master/core/src/main/java/org/openstack4j/openstack/storage/object/internal/ObjectStorageObjectServiceImpl.java#L82 .

**Comment by lsarni** [comment]
If I download without writing to the file it doesn't fail either, it returns a DLPayload that has an HttpResponse saying 404 Not Found.

**Comment by auhlig** [comment]
Ok. I'll mark it as a bug. 
Since you're already familiar with the issue, would you like to contribute and fix it @isarni?
Would be much appreciated.

**Comment by hossein-nasr** [comment]
I think when the request does not read the entity from the response, openstack4j does not handle exceptions at all. I think if any error response received from the server, all connectors should call `HttpExceptionHandler.mapException`  for example at [HttpExecutorServiceImpl.invokeRequest](https://github.com/ContainX/openstack4j/blob/a3558400c9fb1c1ff231864e2548321e8c91ee1c/connectors/httpclient/src/main/java/org/openstack4j/connectors/httpclient/HttpExecutorServiceImpl.java#L64) like what has written in [HttpEntityHandler](https://github.com/ContainX/openstack4j/blob/a3558400c9fb1c1ff231864e2548321e8c91ee1c/core/src/main/java/org/openstack4j/core/transport/HttpEntityHandler.java#L51).
this will throw an exception on any failure.
If it is the preferred outcome, I can fix it.
Otherwise please let me know what was the original idea and how should such an error be handled?
@auhlig

**Comment by hossein-nasr** [comment]
Meanwhile, I did what I said and found out that this is not how openstack4j works because multiple tests had failed because of my change. So error response should not throw an exception which is still weird for me because it questions the existence of ResponseException.
So, I think the `ObjectStorageService` interface should be changed to let the caller access to the error and response headers.
I think we should add some methods which return the `HttpResponse`.
I appreciate if anyone could help me we openstack4j general design about this issue.
